### PR TITLE
Adding a version picker for Synapse docs

### PR DESCRIPTION
**Pull request by Dmytro27Ind**
## 🚀 Adding a version picker for Synapse docs
In this pull request I've added a version picker for the Synapse docs. The problem was that we need to add this version picker for all previous versions of the docs (not just develop). So in addition to adding a version picker, I also wrote a github action to add a version picker for all previous versions of the docs. This action "Add Version Picker (RUN ONCE)" needs to be run manually for develop branch and only once, because then the version picker will be in all versions of the docs and then we don't need to run this action anymore, 

### 📋 Related Issues
This pull request is related to this issue: [#15692](https://github.com/matrix-org/synapse/issues/15692), [#11647](https://github.com/matrix-org/synapse/issues/11647)

### 📝 Instructions
To properly add the version picker, follow these steps:
  1. Merge the pull request.
  2. Manually run the GitHub action **"Add Version Picker (RUN ONCE)"** (docs-add-version-picker.yaml) for develop branch to update the **documentation for previous versions.**

### 📷 Screenshots
As you can see this version picker, can handle all themes.

<div style="display: flex;">
  ver-picker-1
  ver-picker-2
  ver-picker-3
</div>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct.

Signed-off-by: Dmytro Kagirov <dmytro27kagirov@gmail.com>

### CODE REVIEW

**Comment by erikjohnston** [comment]
@Dmytro27Ind Sorry that this hasn't been reviewed yet. We've recently fixed our ability to preview documentation for PRs, so if you could merge in develop then hopefully it'll be easier for us to review (as we'll be able to play around with the real thing)

**Review by erikjohnston — CHANGES_REQUESTED** [review]
(Needs updating)

**Comment by Dmytro27Ind** [comment]
@erikjohnston Hello. I have fixed the conflicts and also added v1.97. Now you can review it

File: changelog.d/16533.doc [hunk 1]
```diff
@@ -0,0 +1 @@
+Added version picker for Synapse documentation. Contributed by @Dmytro27Ind
```
Comment (erikjohnston) [inline changelog.d/16533.doc]:
```suggestion
Added version picker for Synapse documentation. Contributed by @Dmytro27Ind.
```

Comment (erikjohnston) [reply changelog.d/16533.doc]: Ha, we you fixed this already

Comment (Dmytro27Ind) [reply changelog.d/16533.doc]: I've already added it

Comment (erikjohnston) [reply changelog.d/16533.doc]: Github failed  to update :roll_eyes:

**Comment by Dmytro27Ind** [comment]
@erikjohnston It seems that PR Documentation Preview  (https://pr16533--synapse-docs-previews.netlify.app/) only builds the current version of the documentation. You can see the verison picker, but if you change the version, you'll see the not found page:




If you want to be able to play around with it, you can create a fork and build the documentation in your fork.
For example, I built it in my fork and you can see the result here:  https://dmytro27ind.github.io/synapse/v1.95/

**PS v1.96, v1.97 and the latest versions don't work in my github pages documentation preview because I built it over a month ago on v1.95**

File: docs/website_files/version-picker.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+function fetchVersions(dropdown, dropdownMenu) {
+    return new Promise((resolve, reject) => {
+        window.addEventListener("load", () => {
+
+            fetch("https://api.github.com/repos/matrix-org/synapse/git/trees/gh-pages", {
+                cache: "force-cache",
+            }).then(res => 
+                res.json()
+            ).then(resObject => {
+                const excluded = ['dev-docs', 'v1.80.0', 'v1.69.0'];
```
Comment (anoadragon453) [inline docs/website_files/version-picker.js:62]: We have one more version which needs to be excluded:

```suggestion
                const excluded = ['dev-docs', 'v1.91.0', 'v1.80.0', 'v1.69.0'];
```

I don't see any reason why we shouldn't just drop these extra versions from the [gh-pages](https://github.com/matrix-org/synapse/tree/gh-pages) branch. Let's do that in a separate PR though.

File: docs/website_files/version-picker.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            dropdown.querySelector('span').textContent = this.textContent;
+            dropdown.querySelector('input').value = this.getAttribute('id');
+
+            window.location.href = changeVersion(window.location.href, this.textContent);
+        });
+    });
+};
+
+/**
+ * This function fetch the available versions from a GitHub repository
```
Comment (anoadragon453) [inline docs/website_files/version-picker.js:46]:
```suggestion
 * This function fetches the available versions from a GitHub repository
```

File: docs/website_files/version-picker.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    // Put 'develop' and 'latest' at the top
+    if (a === 'develop' || a === 'latest') return -1;
+    if (b === 'develop' || b === 'latest') return 1;
+
+    const versionA = (a.match(/v\d+(\.\d+)+/) || [])[0];
+    const versionB = (b.match(/v\d+(\.\d+)+/) || [])[0];
+
+    // Handle versions without a version number (e.g., 'develop' and 'latest')
+    if (!versionA) return 1;
+    if (!versionB) return -1;
```
Comment (anoadragon453) [inline docs/website_files/version-picker.js:108]: Is this necessary? Aren't we already handling these exceptional versions above?

File: .github/workflows/docs-add-version-picker.yaml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+            # Adding version-picker element to index.hbs
+            awk '/<button id="search-toggle" class="icon-button" type="button" title="Search. \(Shortkey: s\)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">/{
+                print; getline; print; getline; print; getline; print;
+                print "\
+                                  <div class=\"version-picker\">\n\
+                                      <div class=\"dropdown\">\n\
+                                          <div class=\"select\">\n\
+                                              <span></span>\n\
+                                              <i class=\"fa fa-chevron-left\"></i>\n\
```
Comment (anoadragon453) [inline .github/workflows/docs-add-version-picker.yaml:48]: The left chevron here is a bit odd to me. I believe the usual UX for drop-down lists is to have a downwards-facing chevron when the drop-down is closed, and an upwards-facing chevron when the list is expanded.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Thank you very much for tackling this - this is something I've personally wanted to see added since we introduced the documentation website :)

A few small comments below, but on the whole lgtm.

File: .github/workflows/docs-add-version-picker.yaml (hunk 5 above)
Comment (Dmytro27Ind) [reply .github/workflows/docs-add-version-picker.yaml]: Hello. I have fixed it as you can see below:

File: docs/website_files/version-picker.js (hunk 4 above)
Comment (Dmytro27Ind) [reply docs/website_files/version-picker.js]: @anoadragon453 Yes, you're right. I deleted it

File: docs/website_files/version-picker.js (hunk 3 above)
Comment (Dmytro27Ind) [reply docs/website_files/version-picker.js]: Fixed

File: docs/website_files/version-picker.js (hunk 2 above)
Comment (Dmytro27Ind) [reply docs/website_files/version-picker.js]: I deleted v1.91.0 and added the new v1.98 to the github action file (docs-add-version-picker.yaml)

**Review by anoadragon453 — APPROVED** [review]
Thanks for the changes. This now looks good to me!

Merging, and will run the new GitHub action.

**Comment by anoadragon453** [comment]
Hi @Dmytro27Ind, thanks for the run down! We'll (hopefully) be doing another release of Synapse tomorrow, so I think that'd be the easiest path forwards. Otherwise we'd need to cherry-pick https://github.com/matrix-org/synapse/commit/483d22afc39d96b5e25292843806d0c0487e6446 to `master`, but to avoid any potential merge conflicts during the release process, it's probably safer to just wait til tomorrow.

Still, it looks excellent on the lived version'd pages that are up now. Thank you for your work!
